Network: panic due to nil deref when 2 peers configure each other as bootstrap node

They keep reconnecting to each other, ultimately leading to a nil deref.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xccaaff]
goroutine 1323 [running]:
github.com/nuts-foundation/nuts-node/network/p2p.(*managedConnection).receiveMessages.func1(0x0, 0x0, 0xc0009fce10, 0x24, 0xc001a28f60)
	/go/network/p2p/connection.go:153 +0x5f
created by github.com/nuts-foundation/nuts-node/network/p2p.(*managedConnection).receiveMessages
	/go/network/p2p/connection.go:151 +0xb4
```

--- a/network/p2p/connection.go
+++ b/network/p2p/connection.go
@@ -89,8 +89,15 @@ func (conn *managedConnection) send(message *transport.NetworkMessage) error {
 
 func (conn *managedConnection) exchange(messageReceiver messageQueue) {
 	conn.mux.Lock()
+	if conn.messenger == nil {
+		log.Logger().Tracef("Connection already closed (peer=%s)", conn.Peer)
+		conn.mux.Unlock()
+		return
+	}
+	// Use copies of pointers to prevent nil deref when close() is called
 	out := conn.outMessages
-	in := conn.receiveMessages()
+	in := receiveMessages(conn.ID, conn.messenger)
+	messenger := conn.messenger
 	conn.mux.Unlock()
 	for {
 		select {
@@ -99,7 +106,7 @@ func (conn *managedConnection) exchange(messageReceiver messageQueue) {
 				// Connection is closing
 				return
 			}
-			if conn.messenger.Send(message) != nil {
+			if messenger.Send(message) != nil {
 				log.Logger().Warnf("Unable to send message to peer (peer=%s)", conn.Peer)
 			}
 		case message := <-in:
@@ -144,9 +151,7 @@ func (conn *managedConnection) close() {
 
 // receiveMessages spawns a goroutine that receives messages and puts them on the returned channel. The goroutine
 // can only be stopped by closing the underlying gRPC connection.
-func (conn *managedConnection) receiveMessages() chan *PeerMessage {
-	peerID := conn.ID
-	messenger := conn.messenger
+func receiveMessages(peerID PeerID, messenger grpcMessenger) chan *PeerMessage {
 	result := make(chan *PeerMessage, 10)
 	go func() {
 		for {

--- a/network/p2p/connection_test.go
+++ b/network/p2p/connection_test.go
@@ -125,7 +125,7 @@ func Test_connection_receiveMessages(t *testing.T) {
 	// Create connection, start message receiving goroutine
 	conn := newConnection(Peer{}, messenger)
 	defer conn.close()
-	c := conn.receiveMessages()
+	c := receiveMessages(conn.ID, conn.messenger)
 
 	// Wait for numMsg to arrive
 	wg := sync.WaitGroup{}
@@ -187,7 +187,7 @@ func Test_connectionManager_register(t *testing.T) {
 		// Now register second one, disconnect first
 		conn2 := mgr.register(Peer{ID: id, Address: addr + "2"}, nil).(*managedConnection)
 		assert.NotEmpty(t, conn1.closer) // assert first one disconnected
-		assert.Empty(t, conn2.closer) // assert second one connected
+		assert.Empty(t, conn2.closer)    // assert second one connected
 
 		assert.Len(t, mgr.peersByAddr, 1)
 	})

--- a/network/p2p/connection_test.go
+++ b/network/p2p/connection_test.go
@@ -65,6 +65,10 @@ func Test_connection_exchange(t *testing.T) {
 			t.Fatal("expected exchange() to return due to close()")
 		}
 	})
+	t.Run("messenger is nil", func(t *testing.T) {
+		conn := newConnection(Peer{}, nil)
+		conn.exchange(messageQueue{})
+	})
 }
 
 func Test_connection_send(t *testing.T) {

